### PR TITLE
Add default operating system conditionals

### PR DESCRIPTION
**Pull request by xPaw**
This is a basic implementation based on my understanding of how Valve does it (skipping mobile and consoles).

I haven't really tested this — probably need to write tests for this, as this likely doesn't allow consumers to provide their own OS conditionals.

### CODE REVIEW

**Comment by xPaw** [comment]
The only problem I'm thinking of, consumers need to be able to override (or disable) the defaults that library sets.

E.g. if you want WIN32 while running on linux.

**Comment by yaakov-h** [comment]
In that case do we want them to:
1. Remove/clear the current conditions and add new ones?
2. Disable the behavior of having default OS conditionals and then manually adding their new ones?
3. Just say "hey please mimic this OS for me"?

**Comment by xPaw** [comment]
> Remove/clear the current conditions and add new ones?

Isn't this possible with the current code in this PR? Since the conditionals can be cleared.

> Disable the behavior of having default OS conditionals and then manually adding their new ones?

I was thinking of having a boolean option too, but then if we keep readonly conditionals, would need some mechanism to combine OS ones into it.

**Comment by yaakov-h** [comment]
> Isn't this possible with the current code in this PR? Since the conditionals can be cleared.

Yes, but is that the flow/API that we want? i.e. if we're defaulting conditionals, [why] do we still require users to know about platform-specific ones?

**Comment by xPaw** [comment]
> if we're defaulting conditionals, [why] do we still require users to know about platform-specific ones?

If you are on Linux but want to get WIN32 stuff?

**Comment by yaakov-h** [comment]
We could instead have an API something along the lines of `MimicPlatform(OSPlatform)` to simplify it for the consumer but I'm probably overthinking this.

File: ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs [hunk 1]
```diff
@@ -26,6 +23,29 @@ public string[] Conditions
         /// </summary>
         public IIncludedFileLoader FileLoader { get; set; }
 
+        public KVSerializerOptions()
+        {
+            Conditions ??= new List<string>();
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:28]: wait why would this ever be null in the constructor?

Comment (xPaw) [reply ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:28]: are init properties executed after constructor? like so: `new KVSerializerOptions { Conditions = abc }`

Comment (yaakov-h) [reply ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs]: Property assignments like that always run after the constructor.

File: ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs [hunk 2]
```diff
@@ -1,20 +1,17 @@
...
-
         /// <summary>
         /// Gets or sets a list of conditions to use to match conditional values.
         /// </summary>
-        public string[] Conditions
-        {
-            get { return conditions ?? new string[0]; }
-            set { conditions = value; }
-        }
+        public IList<string> Conditions { get; set; } = new List<string>();
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:14]: I don't like having initialization code both inline `{ get; set; } = ` and in the constructor, particularly when one depends on the behaviour/order of the other.

File: ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs [hunk 3]
```diff
@@ -26,6 +23,27 @@ public string[] Conditions
         /// </summary>
         public IIncludedFileLoader FileLoader { get; set; }
 
+        public KVSerializerOptions()
+        {
+            // TODO: In the future we will want to skip this for consoles and mobile devices?
+            Conditions.Add("WIN32");
```
Comment (yaakov-h) [inline ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:29]: Is there a `WIN64` or any equivalent for 64-bit runtimes that we should be including here?

Comment (xPaw) [reply ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:29]: Nope.

File: ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs (hunk 2 above)
Comment (xPaw) [reply ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:14]: Yeah I don't know what's the sanest way of implementing this, feel free to take over.

Comment (yaakov-h) [reply ValveKeyValue/ValveKeyValue/KVSerializerOptions.cs:14]: See #29.

**Pull request merged**